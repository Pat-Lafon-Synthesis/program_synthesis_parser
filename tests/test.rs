use std::{fs::File, io::Read};

use program_synthesis_parser;

macro_rules! make_test {
    ($test_name:tt) => {
        #[test]
        fn $test_name() {
            let parser = program_synthesis_parser::spec::ProblemParser::new();
            let mut file =
                File::open(format!("tests/benchmarks/{}.mls", stringify!($test_name))).unwrap();
            let mut buffer = String::new();
            file.read_to_string(&mut buffer).unwrap();
            insta::assert_debug_snapshot!(parser.parse(&buffer).unwrap());
        }
    };
}

make_test!(bool_always_false);
make_test!(bool_always_true);
make_test!(bool_band);
make_test!(bool_bor);
make_test!(bool_impl);
make_test!(bool_neg);


make_test!(expr_boolean);
make_test!(expr_div);
make_test!(expr_sub);
make_test!(expr);


make_test!(list_append);
make_test!(list_compress);
make_test!(list_concat);
make_test!(list_drop);
make_test!(list_dropeven);
make_test!(list_even_parity);
make_test!(list_filter);
make_test!(list_fold);
make_test!(list_hd);
make_test!(list_inc);
make_test!(list_last);
make_test!(list_last2);
make_test!(list_length);
make_test!(list_make);
make_test!(list_map);
make_test!(list_nth);
make_test!(list_pairwise_swap);
make_test!(list_range);
make_test!(list_rev_append);
make_test!(list_rev_fold);
make_test!(list_rev_snoc);
make_test!(list_rev_tailcall);
make_test!(list_snoc);
make_test!(list_sort_sorted_insert);
make_test!(list_sorted_insert);
make_test!(list_stutter);
make_test!(list_sum);
make_test!(list_take);
make_test!(list_tl);


make_test!(nat_add);
make_test!(nat_iseven);
make_test!(nat_max);
make_test!(nat_mul);
make_test!(nat_pred);
make_test!(nat_sub);


make_test!(tree_balanced);
make_test!(tree_binsert);
make_test!(tree_collect_leaves);
make_test!(tree_count_leaves);
make_test!(tree_count_nodes);
make_test!(tree_height);
make_test!(tree_inorder);
make_test!(tree_lastleft);
make_test!(tree_map);
make_test!(tree_nodes_at_level);
make_test!(tree_notexist);
make_test!(tree_postorder);
make_test!(tree_preorder);
make_test!(tree_sum);