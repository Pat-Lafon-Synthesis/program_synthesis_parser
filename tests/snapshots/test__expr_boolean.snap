---
source: tests/test.rs
expression: parser.parse(&buffer).unwrap()
---
SynthProblem {
    imports: [],
    decls: [
        TypeDecl(
            "bool",
            Type(
                Variant([("True", Type(Tuple([]))), ("False", Type(Tuple([])))]),
            ),
        ),
        TypeDecl(
            "formula",
            Type(
                Variant([("BOOL", Type(Named("bool"))), ("NOT", Type(Named("formula"))), ("ANDALSO", Type(Tuple([Type(Named("formula")), Type(Named("formula"))]))), ("ORELSE", Type(Tuple([Type(Named("formula")), Type(Named("formula"))]))), ("IMPLY", Type(Tuple([Type(Named("formula")), Type(Named("formula"))])))]),
            ),
        ),
        ExpDecl(
            "bnot",
            Exp(
                Fix("bnot", Type(Arrow(Type(Named("bool")), Type(Named("bool")))), Exp(Func(Param { name: "x", ty: Type(Named("bool")) }, Exp(Match(Exp(Var("x")), [(Ctor("False", Wildcard), Exp(CTor("True", Exp(Tuple([]))))), (Ctor("True", Wildcard), Exp(CTor("False", Exp(Tuple([])))))]))))),
            ),
        ),
        ExpDecl(
            "band",
            Exp(
                Fix("band", Type(Arrow(Type(Arrow(Type(Named("bool")), Type(Named("bool")))), Type(Named("bool")))), Exp(Func(Param { name: "x", ty: Type(Named("bool")) }, Exp(Func(Param { name: "y", ty: Type(Named("bool")) }, Exp(Match(Exp(Var("x")), [(Ctor("False", Wildcard), Exp(Var("x"))), (Ctor("True", Wildcard), Exp(Var("y")))]))))))),
            ),
        ),
        ExpDecl(
            "bor",
            Exp(
                Fix("bor", Type(Arrow(Type(Arrow(Type(Named("bool")), Type(Named("bool")))), Type(Named("bool")))), Exp(Func(Param { name: "x", ty: Type(Named("bool")) }, Exp(Func(Param { name: "y", ty: Type(Named("bool")) }, Exp(Match(Exp(Var("x")), [(Ctor("False", Wildcard), Exp(Var("y"))), (Ctor("True", Wildcard), Exp(Var("x")))]))))))),
            ),
        ),
    ],
    synth_type: Type(
        Arrow(Type(Named("formula")), Type(Named("bool"))),
    ),
    spec: UIOEs(
        [
            Example {
                input: [
                    Exp(
                        CTor("BOOL", Exp(Tuple([Exp(CTor("True", Exp(Tuple([]))))]))),
                    ),
                ],
                output: Exp(
                    CTor("True", Exp(Tuple([]))),
                ),
            },
            Example {
                input: [
                    Exp(
                        CTor("BOOL", Exp(Tuple([Exp(CTor("False", Exp(Tuple([]))))]))),
                    ),
                ],
                output: Exp(
                    CTor("False", Exp(Tuple([]))),
                ),
            },
            Example {
                input: [
                    Exp(
                        CTor("NOT", Exp(Tuple([Exp(CTor("BOOL", Exp(Tuple([Exp(CTor("True", Exp(Tuple([]))))]))))]))),
                    ),
                ],
                output: Exp(
                    CTor("False", Exp(Tuple([]))),
                ),
            },
            Example {
                input: [
                    Exp(
                        CTor("NOT", Exp(Tuple([Exp(CTor("BOOL", Exp(Tuple([Exp(CTor("False", Exp(Tuple([]))))]))))]))),
                    ),
                ],
                output: Exp(
                    CTor("True", Exp(Tuple([]))),
                ),
            },
            Example {
                input: [
                    Exp(
                        CTor("ANDALSO", Exp(Tuple([Exp(CTor("BOOL", Exp(Tuple([Exp(CTor("True", Exp(Tuple([]))))])))), Exp(CTor("BOOL", Exp(Tuple([Exp(CTor("False", Exp(Tuple([]))))]))))]))),
                    ),
                ],
                output: Exp(
                    CTor("False", Exp(Tuple([]))),
                ),
            },
            Example {
                input: [
                    Exp(
                        CTor("ANDALSO", Exp(Tuple([Exp(CTor("BOOL", Exp(Tuple([Exp(CTor("True", Exp(Tuple([]))))])))), Exp(CTor("BOOL", Exp(Tuple([Exp(CTor("True", Exp(Tuple([]))))]))))]))),
                    ),
                ],
                output: Exp(
                    CTor("True", Exp(Tuple([]))),
                ),
            },
            Example {
                input: [
                    Exp(
                        CTor("ANDALSO", Exp(Tuple([Exp(CTor("BOOL", Exp(Tuple([Exp(CTor("False", Exp(Tuple([]))))])))), Exp(CTor("BOOL", Exp(Tuple([Exp(CTor("True", Exp(Tuple([]))))]))))]))),
                    ),
                ],
                output: Exp(
                    CTor("False", Exp(Tuple([]))),
                ),
            },
            Example {
                input: [
                    Exp(
                        CTor("ORELSE", Exp(Tuple([Exp(CTor("BOOL", Exp(Tuple([Exp(CTor("True", Exp(Tuple([]))))])))), Exp(CTor("BOOL", Exp(Tuple([Exp(CTor("False", Exp(Tuple([]))))]))))]))),
                    ),
                ],
                output: Exp(
                    CTor("True", Exp(Tuple([]))),
                ),
            },
            Example {
                input: [
                    Exp(
                        CTor("ORELSE", Exp(Tuple([Exp(CTor("BOOL", Exp(Tuple([Exp(CTor("False", Exp(Tuple([]))))])))), Exp(CTor("BOOL", Exp(Tuple([Exp(CTor("False", Exp(Tuple([]))))]))))]))),
                    ),
                ],
                output: Exp(
                    CTor("False", Exp(Tuple([]))),
                ),
            },
            Example {
                input: [
                    Exp(
                        CTor("ORELSE", Exp(Tuple([Exp(CTor("BOOL", Exp(Tuple([Exp(CTor("False", Exp(Tuple([]))))])))), Exp(CTor("BOOL", Exp(Tuple([Exp(CTor("True", Exp(Tuple([]))))]))))]))),
                    ),
                ],
                output: Exp(
                    CTor("True", Exp(Tuple([]))),
                ),
            },
            Example {
                input: [
                    Exp(
                        CTor("IMPLY", Exp(Tuple([Exp(CTor("BOOL", Exp(Tuple([Exp(CTor("True", Exp(Tuple([]))))])))), Exp(CTor("BOOL", Exp(Tuple([Exp(CTor("False", Exp(Tuple([]))))]))))]))),
                    ),
                ],
                output: Exp(
                    CTor("False", Exp(Tuple([]))),
                ),
            },
            Example {
                input: [
                    Exp(
                        CTor("IMPLY", Exp(Tuple([Exp(CTor("BOOL", Exp(Tuple([Exp(CTor("False", Exp(Tuple([]))))])))), Exp(CTor("BOOL", Exp(Tuple([Exp(CTor("False", Exp(Tuple([]))))]))))]))),
                    ),
                ],
                output: Exp(
                    CTor("True", Exp(Tuple([]))),
                ),
            },
            Example {
                input: [
                    Exp(
                        CTor("IMPLY", Exp(Tuple([Exp(CTor("BOOL", Exp(Tuple([Exp(CTor("False", Exp(Tuple([]))))])))), Exp(CTor("BOOL", Exp(Tuple([Exp(CTor("True", Exp(Tuple([]))))]))))]))),
                    ),
                ],
                output: Exp(
                    CTor("True", Exp(Tuple([]))),
                ),
            },
            Example {
                input: [
                    Exp(
                        CTor("IMPLY", Exp(Tuple([Exp(CTor("BOOL", Exp(Tuple([Exp(CTor("True", Exp(Tuple([]))))])))), Exp(CTor("BOOL", Exp(Tuple([Exp(CTor("True", Exp(Tuple([]))))]))))]))),
                    ),
                ],
                output: Exp(
                    CTor("True", Exp(Tuple([]))),
                ),
            },
        ],
    ),
}
